### PR TITLE
Allow specification of a docstring on functions which dispatch on traits (@traitfn)

### PR DESCRIPTION
**Pull request by tehrengruber**
As the name already mentions adds the ability to document "@traitfn" functions like so
```
"some documentation"
@traitfn f(::X) where {X; IsNice{X}} = 1
```

### CODE REVIEW

File: src/SimpleTraits.jl [hunk 1]
```diff
@@ -320,16 +320,16 @@ let
...
         if hasmac
             fn = if oldfn_syntax
-                :(@dummy $fname{$(typs...)}($val, $(strip_kw(args1)...); $(kwargs...)) = ($pushloc; $retsym = $fbody; $poploc; $retsym))
+                :(Base.@__doc__ @dummy $fname{$(typs...)}($val, $(strip_kw(args1)...); $(kwargs...)) = ($pushloc; $retsym = $fbody; $poploc; $retsym))
             else
-                :(@dummy $fname($val, $(strip_kw(args1)...); $(kwargs...)) where {$(typs...)} = ($pushloc; $retsym = $fbody; $poploc; $retsym))
+                :(Base.@__doc__ @dummy $fname($val, $(strip_kw(args1)...); $(kwargs...)) where {$(typs...)} = ($pushloc; $retsym = $fbody; $poploc; $retsym))
             end
-            fn.args[1] = mac # replace @dummy
+            fn.args[findfirst(ex->isexpr(ex) && ex.head==:macrocall, fn.args)].args[1] = mac # replace @dummy
```
Comment (mauro3) [inline src/SimpleTraits.jl]: I think this line does not work anymore.

Comment (tehrengruber) [reply src/SimpleTraits.jl]: Mh, why not? Works fine with julia v1.1 for me. The `findfirst` call should return 3 in any case I am aware of. First element in `fn.args` is the symbol `__doc__` macro, second a `LineNumberNode` and the third the actual function macrocall (see the following exerpt from the ast). I didn't hardcode the 3 just to be safe.

```
3: Expr
      head: Symbol macrocall
      args: Array{Any}((3,))
        1: Symbol @dummy
        2: LineNumberNode
          line: Int64 325
          file: Symbol /home/tehrengruber/.julia/dev/SimpleTraits/src/SimpleTraits.jl
        3: Expr
          head: Symbol =
          args: Array{Any}((2,))
            1: Expr
              head: Symbol where
              args: Array{Any}((2,))
                1: Expr
                2: Symbol K
            2: Expr
              head: Symbol block
              args: Array{Any}((8,))
                1: LineNumberNode
                2: Nothing nothing
                3: LineNumberNode
                4: Expr
                5: LineNumberNode
                6: Nothing nothing
                7: LineNumberNode
                8: Symbol ##376
```

Comment (mauro3) [reply src/SimpleTraits.jl]: Ah, sorry, I was one build behind and was wondering why it didn't work.

**Comment by mauro3** [comment]
Thanks for the PR!  Could you add some tests?  Last time I tried (some time ago now), testing docs was a bit tricky.  Maybe this can help: https://github.com/mauro3/Parameters.jl/blob/808c90cf3b6c6a1cd956f59c2ee3ba0be7b8b96a/test/runtests.jl#L25

**Comment by tehrengruber** [comment]
Sure, I added tests both for a simple function and for one with a prepended macrocall (squashed my commit again).

OT: I wasn't aware that there is an active julia community with meetups in Zuerich, so no. Just joined the the group. Great to know!

**Comment by mauro3** [comment]
Excellent, thanks a lot!

OT: thanks for joining the meetup!  We'll probably meet again after summer.  Also, we're always looking for people to present something.  So email me if you got something.

**Comment by tlnagy** [comment]
Hey @mauro3, can you tag a new release with this feature? It would be nice to have documentable traits.

**Comment by mauro3** [comment]
Done:
https://github.com/JuliaRegistries/General/pull/3875

**Comment by tlnagy** [comment]
Thanks!

**Pull request merged**